packages/labs/src/controllers/updating-controller.ts

```diff
@@ -0,0 +1,101 @@
+ * The complete set of contributors may be found at
+ * http://polymer.github.io/CONTRIBUTORS.txt
+ * Code distributed by Google as part of the polymer project is also
+ * subject to an additional IP rights grant found at
+ * http://polymer.github.io/PATENTS.txt
+ */
+
+import {UpdatingElement} from 'updating-element';
+import {
+  UpdatingMixin,
+  PropertyDeclaration,
+} from 'updating-element/updating-mixin.js';
+export {
+  PropertyDeclaration,
+  PropertyDeclarations,
+} from 'updating-element/updating-mixin.js';
+
+/**
+ * Use this module if you want to create your own base class extending
+ * [[UpdatingController]].
+ * @packageDocumentation
+ */
+
+/**
+ * Base controller class which can interact with an UpdatingElement by hooking
+ * into its lifecycle. The controller can perform tasks when its associated
+ * element is connected or disconnected by implementing the `onConnected` and
+ * `onDisconnected` methods. It can prompt the element to update by calling
+ * `requestUpdate`, and it can perform tasks as the element updates by
+ * implementing `onUpdate` and `onUpdated`. Controllers can contain other
+ * controllers. Controllers have an `element` property which is the element
+ * to which the controller is attached and a `host` property which is either
+ * an element or another controller.
+ * @noInheritDoc
+ */
+
+const UpdatingBase = UpdatingMixin(class {});
+
+export class UpdatingController extends UpdatingBase {
+  /**
+   * UpdatingElement to which this controller is connected.
+   */
+  host: UpdatingElement;
+
+  constructor(host: UpdatingElement) {
+    super();
+    (this.constructor as typeof UpdatingController).finalize();
+    this.host = host;
+    const callbacks = {
+      connectedCallback: this.connectedCallback
+        ? () => this.connectedCallback!()
+        : undefined,
+      disconnectedCallback: this.disconnectedCallback
+        ? () => this.disconnectedCallback!()
+        : undefined,
+      willUpdate: this.willUpdate
+        ? () => this.willUpdate!(this._changedProperties)
+        : undefined,
+      update: this.update
+        ? () => this.update!(this._changedProperties)
+        : undefined,
+      // Note, `didUpdate` is required since it resets `changedProperties`.
+      didUpdate: () => {
+        const changedProperties = this._changedProperties;
+        this._resolveUpdate();
+        this.didUpdate?.(changedProperties);
+      },
+    };
+    host.addController(callbacks);
+  }
+
+  _propertyChanged(
+    name: PropertyKey,
+    oldValue: unknown,
+    _options: PropertyDeclaration
+  ) {
+    if (!this._changedProperties.has(name)) {
+      this._changedProperties.set(name, oldValue);
+    }
+    // When the property changes, provoke an update on the host.
+    this.host.requestUpdate();
+  }
+
+  requestUpdate(
+    name?: PropertyKey,
+    oldValue?: unknown,
+    options?: PropertyDeclaration
+  ) {
+    return name !== undefined
+      ? super.requestUpdate(name, oldValue, options)
```

Comment:
The `super.reqeustUpdate` in `UpdatingMixin` doesn't trigger anything, it just records changed properties, right? Shouldn't we still call `host.requestUpdate` to actually trigger update lifecycle in this case?

Reply:
Oh, the `_propertyChange` override was pretty confusing to me. Can't you get rid of `_propertyChanged` as an override point (always collect changedProperties in the core impl) and make this just do `if (super.reqeustUpdate(...)) { this.host.requestUpdate(); }`?

Reply:
I guess you still need to override `_propertyChange` in `UpdatingElement`, but it seems like `UpdatingController` could be simplified as above.

Reply:
Done.